src/cargo/sources/git/utils.rs

```diff
@@ -565,7 +565,14 @@ pub fn fetch(repo: &git2::Repository, url: &str,
         let mut opts = git2::FetchOptions::new();
         opts.remote_callbacks(cb)
             .download_tags(git2::AutotagOption::All);
-        try!(remote.fetch(&[refspec], Some(&mut opts), None));
-        Ok(())
+
+        match network::with_retry(cargo_config, ||{
+            remote.fetch(&[refspec], Some(&mut opts), None)
+        }){
+            Ok(_) => Ok(()),
+            Err(e) =>{
+                Err(git2::Error::from_str(&format!("{}", e)))
```

Comment:
Aha I see. Hm. So on one hand we need to be a little more proactive here because this will lose error context (`e`'s `Display` implementation only prints itself, not the backtrace of errors). On the other hand this may be difficult to do...

What happens if `with_authentication` requires a closure that returns `CargoResult`?